### ISSUE TITLE
Unbound refuses IPv6 querys after reconnect/restart

### ISSUE THREAD

**Issue by hendrikf87** [opnsense/core#4970]
Hi @fichtner,

I sometimes run into this problem too. I receive a new prefix every night and on the next day queries are "Refused" (Reply code 5). But the ACL file (/var/unbound/access_lists.conf) contains the correct prefixe for each track interface.

However, If I restart, two lines are removed, e.g.:
```
access-control: 2001:12e1:d73b:9200:20d:b9ff:fe57:7758/64 allow
access-control: fd00::20d:b9ff:fe57:7758/64 allow
```
The first line is a address with the prefix from the day before.

Maybe the file is updated correctly (except the two lines) but not loaded by Unbound?

Thanks!

_Originally posted by @Peter0x48 in https://github.com/opnsense/core/issues/3797#issuecomment-827050911_

**Comment by hendrikf87** [comment]
Hi,

sorry for inconvenience, i'm new here and just tried to reopen issue #3797 ;-)

I have the same behaviour as Peter mentioned. Using OPNsense 21.1.5 Unbound will refuse IPv6 querys after restart or reconnect - or generally spoken: when the IPv6 prefix changes. Issue can be replicated...

Assuming this is some sort of "timing problem":
After a restart or connection loss it can take some time (up to 15 minutes) to get an IPv6 prefix delegation. This new prefix is written to the unbound-ACL but unbound will apply this only after restart.

I helped myself by adding a cron job for unbound restart at night - but that's an ugly solution.

Thanks!

**Comment by marjohn56** [comment]
We have completed the dynamic IPv6 rules routines. I'll check  this out and see if it's still a problem here.

**Comment by marjohn56** [comment]
Seems so.. looking into it.

**Comment by fichtner** [comment]
Essentially we need to use unbound-control to adjust ACL... we cannot afford to bounce the process along with its cache for each ACL change in the system.

**Comment by fichtner** [comment]
PS: This ticket is a duplicate of #4466

**Comment by marjohn56** [comment]
I've been playing.. as usual. I've added an unbound_generate_config() and unbound reload to newwanipv6, seems to work but not 100% convinced yet.

**Comment by fichtner** [comment]
We simply end up reloading unbound on all events all the time, might as well use a cron job every other minute ;)

**Comment by marjohn56** [comment]
Well using newwanipv6 means it only reloads it when dhcp6c calls the script, I would have thought that was a better option.

**Comment by fichtner** [comment]
It is already called from this place.

https://github.com/opnsense/core/blob/05d3224e7017940f43c307420e44c8cea5034b80/src/etc/rc.newwanipv6#L146

https://github.com/opnsense/core/blob/05d3224e7017940f43c307420e44c8cea5034b80/src/etc/inc/plugins.inc.d/unbound.inc#L47

**Comment by marjohn56** [comment]
It is... but the ACL is not *always* updated.. hmm odd.

**Comment by hendrikf87** [comment]
Hi,

do you guys changed something in 21.7.x?

Just checked my Windows Ipconfig, i get an IPv6 address but no IPv6 DNS Server anymore. Checked that with two Windows Clients Build 1909 and 20H2. If i disable IPv4 Stack the Windows Client gets an IPv6 DNS Server instantly.

iOS Devices get an IPv4 DNS as Primary and an IPv6 DNS as Secondary.

For me that's better than before (IPv6 DNS as primary on Windows Clients). IPv4 DNS could also resolve AAAA records and is working without interruption.

Just trying to understand what happened here ;-)

Best regards
Hendrik

**Comment by fichtner** [comment]
21.7 changed the reload behaviour to always reload on interface event since we have the ability to retain the cache now. I can't speak for not handing out DNS server on IPv6. Seems unrelated.
